### PR TITLE
inventory viewer: add item grouping and free slots display

### LINKED ISSUE #2771: Count of how many of a certain item you have in your inventory
Would be useful when using superglassmake and buying from Charter Crewmembers on an ironman.

### PR DESCRIPTION
**Pull request by Hydrox6**
Closes #2771

Adds an option to have the inventory viewer group the same items together.

Now with an option to show the number of empty slots remaining in your inventory

java_2019-03-02_01-03-42
java_2019-03-02_01-04-17

### CODE REVIEW

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 1]
```diff
@@ -105,4 +146,9 @@ private BufferedImage getImage(Item item)
 		ItemComposition itemComposition = itemManager.getItemComposition(item.getId());
 		return itemManager.getImage(item.getId(), item.getQuantity(), itemComposition.isStackable());
 	}
+
+	private BufferedImage getImage(int id, int quantity)
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: what is the point of this method.

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerConfig.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+public interface InventoryViewerConfig extends Config
+{
+	@ConfigItem(
+		keyName = "viewerMode",
+		name = "Mode",
+		description = "The mode to display the inventory viewer with"
+	)
+	default InventoryViewerMode viewerMode()
+	{
+		return InventoryViewerMode.FULL;
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerConfig.java]: Use boolean

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java (hunk 1 above)
Comment (Hydrox6) [reply runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: This used to have more in it. Thanks for spotting it

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 3]
```diff
@@ -77,24 +86,42 @@ public Dimension render(Graphics2D graphics)
...
 					}
 				}
+
+				// put a placeholder image so each item is aligned properly and the panel is not resized
+				panelComponent.getChildren().add(PLACEHOLDER_IMAGE);
 			}
+		}
+		else
+		{
+			final Map<Integer, Integer> totals = Arrays.stream(items)
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: This else block can be in an if statement before the above block with early return

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 4]
```diff
@@ -39,7 +43,10 @@
 import net.runelite.client.ui.overlay.OverlayPosition;
 import net.runelite.client.ui.overlay.components.ImageComponent;
 import net.runelite.client.ui.overlay.components.PanelComponent;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.summingInt;
 
+@Slf4j
```
Comment (raiyni) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: This is unused?

Comment (Hydrox6) [reply runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: Oops, ty

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 5]
```diff
@@ -77,6 +84,23 @@ public Dimension render(Graphics2D graphics)
...
+
+			for (Map.Entry<Integer, Integer> cursor : totals.entrySet())
+			{
+				final BufferedImage image = itemManager.getImage(cursor.getKey(), cursor.getValue(), true);
+				if (image != null)
+				{
+					panelComponent.getChildren().add(new ImageComponent(image));
+				}
+			}
+			return panelComponent.render(graphics);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: Add empty line above

**Comment by Hydrox6** [comment]
Added option to show free slots remaining. Let me know if I've done anything really badly wrong

**Comment by ShaunDreclin** [comment]
How are the items sorted? I would think it'd be better to do it by the first time each item appears in the inventory

**Comment by deathbeam** [comment]
Its currently using set for w/e reason so I dont think its order makes any sense anymore

**Comment by raiyni** [comment]
Natural order of multiset

**Comment by Hydrox6** [comment]
Updated and refactored the code, and made it use the old method of gathering the items.

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 6]
```diff
@@ -80,49 +112,78 @@ public Dimension render(Graphics2D graphics)
...
+						}
 					}
 				}
+
+				// put a placeholder image so each item is aligned properly and the panel is not resized
+				inventoryComponent.getChildren().add(PLACEHOLDER_IMAGE);
 			}
+			wrapperWidth = 4 * (Constants.ITEM_SPRITE_WIDTH + 6);
+		}
+		// If the inventoryComponent is empty, the overlay NPEs during rendering
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: empty line above

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 7]
```diff
@@ -80,49 +112,78 @@ public Dimension render(Graphics2D graphics)
...
+							inventoryComponent.getChildren().add(new ImageComponent(image));
+							continue;
+						}
 					}
 				}
+
+				// put a placeholder image so each item is aligned properly and the panel is not resized
+				inventoryComponent.getChildren().add(PLACEHOLDER_IMAGE);
 			}
+			wrapperWidth = 4 * (Constants.ITEM_SPRITE_WIDTH + 6);
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: empty line above

**Comment by Hydrox6** [comment]
Moved the grouped placeholder stuff to the first commit, since it makes more sense to be in there anyway

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 8]
```diff
@@ -78,6 +101,31 @@ public Dimension render(Graphics2D graphics)
...
+				if (image != null)
+				{
+					panelComponent.getChildren().add(new ImageComponent(image));
+				}
+			}
+
+			// Add a placeholder if the inventory is empty, so the overlay can still be easily seen
+			if (totals.entrySet().size() == 0)
+			{
+				panelComponent.getChildren().add(inventoryIconSprite);
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: I feel this icon should be somewhat transparent, similar to a bank placeholder, so it is as clear as possible that there are no items present.

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 9]
```diff
@@ -50,18 +57,34 @@
...
+		{
+			final BufferedImage img = new BufferedImage(Constants.ITEM_SPRITE_WIDTH, Constants.ITEM_SPRITE_HEIGHT, BufferedImage.TYPE_INT_ARGB);
+			final int centeredY = Constants.ITEM_SPRITE_HEIGHT / 2 - sprite.getHeight() / 2;
+
+			final Graphics2D g = img.createGraphics();
+			g.drawImage(sprite, 0, centeredY, null);
+			g.dispose();
+
+			inventoryIconSprite = new ImageComponent(img);
+		});
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]:
```java
spriteManager.getSpriteAsync(SpriteID.TAB_INVENTORY, 0, sprite ->
{
	inventoryIconSprite = ImageUtil.resizeCanvas(sprite, Constants.IMAGE_SPRITE_WIDTH, Constants.IMAGE_SPRITE_HEIGHT);
});
```

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 10]
```diff
@@ -45,31 +49,59 @@
...
+			0,
+			ComponentConstants.STANDARD_BORDER));
+
+		wrapperComponent.setOrientation(ComponentOrientation.VERTICAL);
+		wrapperComponent.setWrapping(2);
+		wrapperComponent.setBorder(new Rectangle(
+			ComponentConstants.STANDARD_BORDER,
+			ComponentConstants.STANDARD_BORDER,
+			ComponentConstants.STANDARD_BORDER,
+			ComponentConstants.STANDARD_BORDER));
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: This is the default border that PanelComponent sets, so it's not necessary to define it as such here.

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 11]
```diff
@@ -45,31 +49,59 @@
...
+		inventoryComponent.setOrientation(ComponentOrientation.HORIZONTAL);
+		inventoryComponent.setBackgroundColor(null);
+		inventoryComponent.setBorder(new Rectangle(
+			ComponentConstants.STANDARD_BORDER,
+			0,
+			0,
+			ComponentConstants.STANDARD_BORDER));
+
+		wrapperComponent.setOrientation(ComponentOrientation.VERTICAL);
+		wrapperComponent.setWrapping(2);
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: The wrapper should not be doing any wrapping should it?

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 12]
```diff
@@ -92,61 +124,100 @@ public Dimension render(Graphics2D graphics)
...
+			if (groupedItems.entrySet().size() == 0)
 			{
-				panelComponent.getChildren().add(inventoryIconSprite);
+				inventoryComponent.getChildren().add(inventoryIconSprite);
+				wrapperWidth = WRAPPER_MINIMUM_WIDTH;
+			}
+			else
+			{
+				final int widthItems = Math.min(INVENTORY_WIDTH, groupedItems.entrySet().size());
+				wrapperWidth = widthItems * Constants.ITEM_SPRITE_WIDTH + (widthItems + 1) * SPRITE_PADDING.x;
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: This doesn't use the same formula as is used for the non-grouped inventory. What's below is:

```java
wrapperWidth = INVENTORY_WIDTH * Constants.ITEM_SPRITE_WIDTH + (INVENTORY_WIDTH - 1) * SPRITE_PADDING.x;
```

Specifically, **`(INVENTORY_WIDTH - 1)`**` * SPRITE_PADDING.x` compared to **`(widthItems + 1)`**` * SPRITE_PADDING.x`. Why are these different?

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 13]
```diff
@@ -92,61 +124,100 @@ public Dimension render(Graphics2D graphics)
...
 				}
+
+				// put a placeholder image so each item is aligned properly and the panel is not resized
+				inventoryComponent.getChildren().add(PLACEHOLDER_IMAGE);
 			}
+			
+			wrapperWidth = INVENTORY_WIDTH * Constants.ITEM_SPRITE_WIDTH + (INVENTORY_WIDTH - 1) * SPRITE_PADDING.x;
+		}
+
+		// If the inventoryComponent is empty, the overlay NPEs during rendering
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: Unrelated to your code, but is this a bug? Seems like that should not happen.

File: runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java [hunk 14]
```diff
@@ -27,78 +28,196 @@
...
 
-		final Item[] items = itemContainer.getItems();
+		if (config.groupItems())
+		{
+			if (groupedItems == null)
+			{
+				// Store in a LinkedHashMap to preserve inventory ordering
+				groupedItems = Arrays.stream(items)
+					.filter(p -> p.getId() != -1)
+					.collect(groupingBy(Item::getId, LinkedHashMap::new, summingInt(Item::getQuantity)));
```
Comment (Nightfirecat) [inline runelite-client/src/main/java/net/runelite/client/plugins/inventoryviewer/InventoryViewerOverlay.java]: Does this order the grouped items in the same order they appear in your inventory? From the screenshots provided on this PR, it seems like it does not.

**Comment by Hydrox6** [comment]
Added to the hub with runelite/plugin-hub#189

**Pull request closed without merging**